aptos-move/framework/aptos-framework/sources/storage_deposit.move

```diff
@@ -0,0 +1,103 @@
+    const EGLOBAL_STORAGE_DEPOSIT: u64 = 0;
+
+    struct GlobalStorageDeposit has key, store {
+        deposit: AggregatableCoin<AptosCoin>,
+        slot_deposit_event: EventHandle<SlotDepositEvent>,
+        excess_bytes_penalty_event: EventHandle<ExcessBytesPenaltyEvent>,
+        slot_refund_event: EventHandle<SlotRefundEvent>,
+    }
+
+    struct SlotDepositEvent has drop, store {
+        payer: address,
+        amount: u64,
+    }
+
+    struct ExcessBytesPenaltyEvent has drop, store {
+        payer: address,
+        amount: u64,
+    }
+
+    struct SlotRefundEvent has drop, store {
+        payee: address,
+        amount: u64,
+    }
+
+    struct DepositEntry {
+        account: address,
+        amount: u64,
+    }
+
+    public entry fun initialize(aptos_framework: &signer) {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<GlobalStorageDeposit>(@aptos_framework),
+            error::already_exists(EGLOBAL_STORAGE_DEPOSIT)
+        );
+
+        let global_storage_deposit = GlobalStorageDeposit {
+            // FIXME(aldenhu): needs the limit to be u128 (u64 implied)
+            deposit: coin::initialize_aggregatable_coin<AptosCoin>(aptos_framework),
+            slot_deposit_event: new_event_handle<SlotDepositEvent>(aptos_framework),
+            excess_bytes_penalty_event: new_event_handle<ExcessBytesPenaltyEvent>(aptos_framework),
+            slot_refund_event: new_event_handle<SlotRefundEvent>(aptos_framework),
+        };
+
+        move_to<GlobalStorageDeposit>(aptos_framework, global_storage_deposit);
+    }
+
+    public(friend) fun charge_and_refund(
+        slot_charges: vector<DepositEntry>,
+        slot_refunds: vector<DepositEntry>,
+        excess_bytes_penalties: vector<DepositEntry>,
+    ) acquires GlobalStorageDeposit {
+        assert!(
+            exists<GlobalStorageDeposit>(@aptos_framework),
+            error::not_found(EGLOBAL_STORAGE_DEPOSIT)
+        );
+        let global_storage_deposit = borrow_global<GlobalStorageDeposit>(@aptos_framework);
+
+        let i = 0;
+        let len = vector::length(&slot_charges);
+        while (i <= len) {
+            let entry = vector::borrow(&slot_charges, i);
+            coin::collect_into_aggregatable_coin<AptosCoin>(entry.account, entry.amount, &mut global_storage_deposit.deposit);
+            // FIXME(aldenhu): central events kills concurrency, probably need to augment Account with these events
+            // TODO: emit event
+            i = i + 1;
+        };
+
+        let i = 0;
+        let len = vector::length(&slot_refunds);
+        while (i <= len) {
+            let entry = vector::borrow(&slot_charges, i);
+            let coin = coin::extract_from_aggregatable_coin(&mut global_storage_deposit.deposit, entry.amount);
+            coin::deposit(entry.account, coin);
+            // FIXME(aldenhu): central events kills concurrency, probably need to augment Account with these events
```

Comment:
@davidiw ..

Reply:
Considering that there'd be no way to know why you get refunds, I'm not sure a way around this.

Reply:
Now that it's part of the gas, I'm thinking being obscure (just like other gas charges) is fine 🤡

Reply:
Oh.. the refund won't be deducted from `gas_used` though..

Reply:
With the movement toward objects, we could consider explicit user-owned objects for reimbursements, but there's a whole host of other issues that need to be addressed, such as opt-in, what to do if the user has only one -- do we have a framework to maximize parallelization..